backend/src/schema/resolvers/images/images.spec.js

```diff
@@ -0,0 +1,344 @@
+import { deleteImage, mergeImage } from './images'
+import { getNeode, getDriver } from '../../../db/neo4j'
+import Factory, { cleanDatabase } from '../../../db/factories'
+import { UserInputError } from 'apollo-server'
+
+const driver = getDriver()
+const neode = getNeode()
+const uuid = '[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}'
+let uploadCallback
+let deleteCallback
+
+beforeEach(async () => {
+  await cleanDatabase()
+  uploadCallback = jest.fn(({ destination }) => destination)
+  deleteCallback = jest.fn()
+})
+
+describe('deleteImage', () => {
+  describe('given a resource with an image', () => {
+    let user
+    beforeEach(async () => {
+      user = await Factory.build(
+        'user',
+        {},
+        {
+          avatar: Factory.build('image', {
+            url: '/some/avatar/url/',
+            alt: 'This is the avatar image of a user',
+          }),
+        },
+      )
+      user = await user.toJson()
+    })
+
+    it('soft deletes `Image` node', async () => {
+      await expect(neode.all('Image')).resolves.toHaveLength(1)
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      await expect(neode.all('Image')).resolves.toHaveLength(0)
+    })
+
+    it('calls deleteCallback', async () => {
+      user = await Factory.build('user')
+      user = await user.toJson()
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      expect(deleteCallback).toHaveBeenCalled()
+    })
+
+    describe('given a transaction parameter', () => {
+      it('executes cypher statements within the transaction', async () => {
+        const session = driver.session()
+        let someString
+        try {
+          someString = await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            const txResult = await transaction.run('RETURN "Hello" as result')
+            const [result] = txResult.records.map(record => record.get('result'))
+            return result
+          })
+        } finally {
+          session.close()
+        }
+        await expect(neode.all('Image')).resolves.toHaveLength(0)
+        await expect(someString).toEqual('Hello')
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-25T16:15:28Z" title="Tuesday, February 25th 2020, 5:15:28 pm +01:00">Feb 25, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,347 @@
+import { deleteImage, mergeImage } from './images'
+import { getNeode, getDriver } from '../../../db/neo4j'
+import Factory, { cleanDatabase } from '../../../db/factories'
+import { UserInputError } from 'apollo-server'
+
+const driver = getDriver()
+const neode = getNeode()
+const uuid = '[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}'
+let uploadCallback
+let deleteCallback
+
+beforeEach(async () => {
+  await cleanDatabase()
+  uploadCallback = jest.fn(({ destination }) => destination)
+  deleteCallback = jest.fn()
+})
+
+describe('deleteImage', () => {
+  describe('give a resource with an image', () => {
+    let user
+    beforeEach(async () => {
+      user = await Factory.build(
+        'user',
+        {},
+        {
+          avatar: Factory.build('image', {
+            url: '/some/avatar/url/',
+            alt: 'This is the avatar image of a user',
+          }),
+        },
+      )
+      user = await user.toJson()
+    })
+
+    it('soft deletes `Image` node', async () => {
+      await expect(neode.all('Image')).resolves.toHaveLength(1)
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      await expect(neode.all('Image')).resolves.toHaveLength(0)
+    })
+
+    it('calls deleteCallback', async () => {
+      user = await Factory.build('user')
+      user = await user.toJson()
+      await deleteImage(user, 'AVATAR_IMAGE', { deleteCallback })
+      expect(deleteCallback).toHaveBeenCalled()
+    })
+
+    describe('given a transaction parameter', () => {
+      it('executes cypher statements within the transaction', async () => {
+        const session = driver.session()
+        let someString
+        try {
+          someString = await session.writeTransaction(async transaction => {
+            await deleteImage(user, 'AVATAR_IMAGE', {
+              deleteCallback,
+              transaction,
+            })
+            const txResult = await transaction.run('RETURN "Hello" as result')
+            const [result] = txResult.records.map(record => record.get('result'))
+            return result
+          })
+        } finally {
+          session.close()
+        }
+        await expect(neode.all('Image')).resolves.toHaveLength(0)
+        await expect(someString).toEqual('Hello')
```


I don't get this test, the third parameter is the `opts`, which you can pass in a transaction, and then it will use that to run the cypher to check if an image exists, and if so deletes the relationship, and then delete the image on the filesystem, or spaces, correct?
what are you intending to test here? `executes cypher statements within the transaction`, but you are not testing that it is run in the `deleteImage` function, you are running your own `transaction`, `transaction.run('RETURN "Hello" as result')`, then checking that this is saved to a variable `someString`? 
Maybe, I don't understand what you are trying to test here, but to me it looks like you are testing that the `neo4j-javascript-driver` works as expected... :woman_shrugging:

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-02T16:28:50Z" title="Monday, March 2nd 2020, 5:28:50 pm +01:00">Mar 2, 2020</time>_

---


Kind of. I would argue that it is the happy path for the actual test case where I introduce a bug which should roll back the entire transaction.